rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/LocalInfrastructure.java

```diff
@@ -389,22 +327,49 @@ protected void initializePersistedInfraVariables() {
         this.persistedInfraVariables.put(LAST_NODE_STARTED_INDEX_KEY, 0);
     }
 
-    private void removeAcquiredNodesAndShutDownIfNeeded() {
-        int remainingNodesCount = decrementAndGetNumberOfAcquiredNodesWithLockAndPersist();
-        // if there is no remaining node, kill the JVM process
-        if (remainingNodesCount == 0) {
-            shutDown();
+    private void removeNodeAndShutdownProcessIfNeeded(String nodeUrl) {
+
+        // Update handle & acquire/lost nodes persistent counters
+        decrementNumberOfHandledNodesWithLockAndPersist();
+        if (nodeSource.getNodeInDeployingOrLostNodes(nodeUrl) == null) {
+            decrementNumberOfLostNodesWithLockAndPersist();
+        } else {
+            decrementNumberOfAcquiredNodesWithLockAndPersist();
+        }
+
+        // Delete the process if it doesn't have remaining node
+        Iterator<Map.Entry<ProcessExecutor, List<String>>> processIterator = processExecutors.entrySet().iterator();
+        while (processIterator.hasNext()) {
+            Map.Entry<ProcessExecutor, List<String>> processExecutor = processIterator.next();
+            if (processExecutor.getValue().contains(nodeUrl)) {
+                processExecutor.getValue().remove(nodeUrl);
+                // if there is no remaining node, kill the JVM process
+                if (processExecutor.getValue().isEmpty()) {
+                    if (processExecutor.getKey() != null) {
+                        processExecutor.getKey().killProcess();
+                    }
+                    processIterator.remove();
+                }
+            }
         }
     }
 
     // Below are wrapper methods around the runtime variables map
 
+    private int getNumberOfHandledNodesWithLock() {
+        return getPersistedInfraVariable(() -> (int) this.persistedInfraVariables.get(NB_HANDLED_NODES_KEY));
+    }
+
     private int getNumberOfAcquiredNodesWithLock() {
         return getPersistedInfraVariable(() -> (int) this.persistedInfraVariables.get(NB_ACQUIRED_NODES_KEY));
     }
 
-    private int getNumberOfLostNodesWithLock() {
-        return getPersistedInfraVariable(() -> (int) this.persistedInfraVariables.get(NB_LOST_NODES_KEY));
+    private void increaseNumberOfHandledNodesWithLockAndPersist(final int additionalNumberOfNodes) {
+        setPersistedInfraVariable((PersistedInfraVariablesHandler<Void>) () -> {
+            int updated = (int) this.persistedInfraVariables.get(NB_HANDLED_NODES_KEY) + additionalNumberOfNodes;
+            this.persistedInfraVariables.put(NB_HANDLED_NODES_KEY, updated);
+            return null;
```

Comment:
is it fine that you return null?

Reply:
Yes as we don't use the return value, see `PersistedInfraVariablesHandler<Void>`. Changed in fe1041da4cb996c6841786fc000dfb120c836238 as it is confusing.